snowflake_blueprints/upload_file.py

```diff
@@ -65,23 +59,209 @@ def get_args():
     return args
 
 
-def upload_data(source_full_path, table_name, insert_method, db_connection):
+def create_table(source_full_path, table_name, insert_method, db_connection):
+    """
+    Creates a table by looking at the schema of the first 10k rows and only loading the header row.
+    Used by the new PUT method because you can't PUT or COPY INTO if the table doesn't exist beforehand.
+    """
     try:
+        chunksize = 10000
         for index, chunk in enumerate(
-                pd.read_csv(source_full_path, chunksize=10000)):
-
-            if insert_method == 'replace' and index > 0:
-                # First chunk replaces the table, the following chunks
-                # append to the end.
-                insert_method = 'append'
-
-            chunk.to_sql(
+                pd.read_csv(source_full_path, chunksize=chunksize)):
+            chunk.head(0).to_sql(
                 table_name,
                 con=db_connection,
-                index=False,
-                if_exists=insert_method,
-                method='multi',
-                chunksize=10000)
+                if_exists="fail",
+                index=False)
+            # prevents a loop that was necessary for loading only a small chunk
+            # of the data in.
+            break
+        if insert_method == 'append':
+            print(f'Created table {table_name} because it did not exist.')
+        elif insert_method == 'replace':
+            print(f'Created a new table {table_name}.')
+    except BaseException as e:
+        if 'already exists' in str(e):
+            pass
+        else:
+            print(e)
+
+
+def convert_to_parquet(source_full_path, table_name):
+    """
+    Converts a given CSV to multiple Parquet for uploading.
+    Uses fastest tested method with Dask and gzip (Snowflake recommendation)
+    Parquet files allows for column name mapping.
+    """
+    parquet_path = f'./tmp/{table_name}'
+    shipyard.files.create_folder_if_dne(parquet_path)
+    df = dd.read_csv(source_full_path)
+    df.columns = map(lambda x: str(x).upper(), df.columns)
+    df.to_parquet(parquet_path, compression='gzip', write_index=False)
+    return parquet_path
+
+
+def execute_put_command(db_connection, file_path, table_name, results_dict):
+    """
+    Execute the PUT command against Snowflake and store the results.
+    """
+    put = db_connection.execute(f'PUT file://{file_path}/* @%"{table_name}"')
+    for item in put:
+        # These are guesses. The documentation doesn't specify.
+        put_results = {
+            "input_file_name": item[0],
+            "uploaded_file_name": item[1],
+            "input_bytes": item[2],
+            "uploaded_bytes": item[3],
+            "input_file_type": item[4],
+            "uploaded_file_type": item[5],
+            "status": item[6],
+            "unknown": item[7]}
```

Comment:
> These are guesses. 
Are you sure the number of columns exist? I.e. can we be sure that item[7] won't ever cause an out of bounds exception because there are less than 8 items?

Reply:
I cannot be sure that `item[7]` won't every cause an out of bounds exception. In all testing instances, it returned a blank string. I figured capturing it was the better practice, but I think removing it is probably the right call.